misc-scripts/xref_mapping/XrefParser/UniProtParser/Extractor.pm

```diff
@@ -0,0 +1,645 @@
+
+  Readonly my $id_status_field
+    => qr{
+           (?: Unreviewed )
+         | (?: Reviewed )
+       }msx;
+
+  # These is only one ID line
+  my $id_line = $self->{'record'}->{'ID'}->[0];
+  my ( $entry_status )
+    = ( $id_line =~ m{
+                       \A
+
+                       # UniProt name
+                       [0-9A-Z_]+
+
+                       \s+
+
+                       ( $id_status_field )
+                       \s*
+                       ;
+                   }msx );
+  if ( ! defined $entry_status ) {
+    croak "Invalid entry status in:\n\t$id_line";
+  }
+
+  # Likewise, these is only one PE line
+  my $pe_line = $self->{'record'}->{'PE'}->[0];
+  my ( $evidence_level )
+    = ( $pe_line =~ m{
+                       \A
+
+                       ( [1-4] )
+                       \s*
+                       :
+                   }msx );
+  if ( ! defined $evidence_level ) {
+    croak "Invalid protein evidence level in:\n\t$pe_line";
+  }
+
+  return {
+          'status'         => $entry_status,
+          'evidence_level' => $evidence_level,
+        };
+}
+
+
+# Parse the sequence ('  ') fields of the current record and produce
+# its polypeptide sequence. as a continuous string i.e. without
+# decorative whitespace.
+sub _get_sequence {
+  my ( $self ) = @_;
+
+  my $sequence_fields = $self->{'record'}->{ q{  } };
+
+  # Concatenate the sequence into a single continuous string. We do
+  # not expect to see more than whitespace at a time so instead of
+  # trying to match as long a string of them as possible in order to
+  # minimise the number of independent substitutions, we always match
+  # on one in order to avoid unnecessary backtracking.
+  # Note that we use non-destructive substitution.
+  my $sequence = ( join( q{}, @{ $sequence_fields } ) =~ s{ \s }{}grmsx );
+
+  # We could in principle directly return substitution result but then
+  # we would have to make sure we always call get_sequence() in scalar
+  # context. Safer to simply return a scalar instead.
+  return $sequence;
+}
+
+
+
+# Parse the OX field of the current record and produce a list of
+# database_qualifier/taxon_code pairs.
+sub _get_taxon_codes {
+  my ( $self ) = @_;
+
+  # There is only one OX line
+  my $ox_line = $self->{'record'}->{'OX'}->[0];
+
+  # On the one hand, according to UniProt-KB User Manual from October
+  # 2018 there should only be a single taxon code per OX line and the
+  # current SwissProt data file doesn't contain any records which
+  # violate this. On the other hand we haven't checked this in the
+  # TrEMBL file because of its size and the old UniProtParser did have
+  # support for synonyms (albeit using different syntax -
+  # "db=id1, id2, ...;" rather than "db1=id1; db2=id2; ...").
+  # The code below assumes there might be multiple entries present, if
+  # you want to force it to only ever look for one simply drop the /g
+  # modifier from the regex match.
+
+  Readonly my $taxon_db_entry
```

Comment:
As per my earlier comment about seemingly defining in every iteration.

Reply:
Fixed as well, along with all the other occurences of regex constants in Extractor.